<a href="https://colab.research.google.com/github/Arhin-Eben/Machine-learning-with-python/blob/master/Grid_Search_W3schools.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Grid Search**

In [ ]:
from sklearn import datasets
iris = datasets.load_iris()


In [ ]:
#independent variables X and a dependant variable y.
X = iris['data']
y = iris['target']

In [ ]:
#load the logistic model for classifying the iris flowers
from sklearn.linear_model import LogisticRegression

#Creating the model, setting max_iter to a higher value to ensure that the model finds a result
logit = LogisticRegression(max_iter=1000)


In [ ]:
#fit the model to the data
print(logit.fit(X,y))

LogisticRegression(max_iter=1000)


In [ ]:
print(logit.score(X,y))

0.9733333333333334


With the default setting of C = 1, we achieved a score of 0.973

In [ ]:
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

''

Let's see if we can do any better by implementing a grid search with difference values of 0.973.

In [18]:
#Implementing Grid Search
#Since the default value for C is 1, we will set a range of values surrounding it.
C = [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]

#for loop to change out the values of C and evaluate the model with each change
scores = []

for choice in C:
  logit.set_params(C=choice)
  logit.fit(X,y)
  scores.append(logit.score(X,y))

#With the scores stored in a list, we can evaluate what the best choice of C is
  print(scores)

[0.9666666666666667]
[0.9666666666666667, 0.9666666666666667]
[0.9666666666666667, 0.9666666666666667, 0.9733333333333334]
[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334]
[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98]
[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98, 0.98]
[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98, 0.98, 0.9866666666666667]
[0.9666666666666667, 0.9666666666666667, 0.9733333333333334, 0.9733333333333334, 0.98, 0.98, 0.9866666666666667, 0.9866666666666667]


We can see that the lower values of C performed worse than
the base parameter of 1.

However, as we increased the value of C to 1.75 the model experienced increased accuracy.


It seems that increasing C beyond this amount does not
help increase model accuracy.